src/hooks/useNodeViewPortals.tsx

```diff
@@ -0,0 +1,64 @@
+import React, { ReactPortal, useCallback, useState } from "react";
+import { createPortal } from "react-dom";
+
+import {
+  PORTAL_REGISTRY_ROOT_KEY,
+  PortalRegistryContext,
+  PortalRegistryKey,
+} from "../contexts/PortalRegistryContext.js";
+import { RegisterElement } from "../nodeViews/createReactNodeViewConstructor.js";
+
+type Props = {
+  portals: Record<PortalRegistryKey, ReactPortal[]>;
+};
+
+function NodeViews({ portals }: Props) {
+  const rootPortals = portals[PORTAL_REGISTRY_ROOT_KEY];
+
+  return (
+    <PortalRegistryContext.Provider value={portals}>
+      {rootPortals}
+    </PortalRegistryContext.Provider>
+  );
+}
+
+/**
+ * Provides an element rendering the node view portals and a
+ * callback for registering new portals.
+ *
+ * The `registerPortal` callback is meant to be passed to
+ * `createNodeViewConstructor` as the `registerElement` argument. The
+ * `portals` element should be passed as children to the `ProseMirror`
+ * component.
+ */
+export function useNodeViewPortals() {
+  const [portals, setPortals] = useState(
+    {} as Record<PortalRegistryKey, ReactPortal[]>
+  );
+
+  const registerPortal: RegisterElement = useCallback(
+    (registrationKey, child, container, key) => {
+      const portal = createPortal(child, container, key);
+      setPortals((oldPortals) => {
+        return {
+          ...oldPortals,
+          [registrationKey]: (oldPortals[registrationKey] ?? []).concat(portal),
+        };
+      });
+
+      return () => {
+        setPortals((oldPortals) => {
+          return {
+            ...oldPortals,
+            [registrationKey]: (oldPortals[registrationKey] ?? []).filter(
+              (p) => p !== portal
+            ),
+          };
+        });
+      };
+    },
+    []
+  );
+
+  return { portals: <NodeViews portals={portals} />, registerPortal };
```

Comment:
Should we rename this key? It's not a list of portals, but a single `NodeViews` element now, yeah?

Reply:
Yeah, I just took all of this and collapsed it into `useNodeViews` to avoid figuring out what to call this :) haha